site/specification/ORCv1.md

```diff
@@ -369,6 +397,194 @@ message Metadata {
+enables access that column in that file. The local keys are encrypted
+by the Hadoop or Ranger Key Management Server (KMS). The encrypted
+local keys are stored in the file footer's StripeInformation.
+
+```
+enum KeyProviderKind {
+  UNKNOWN = 0;
+  HADOOP = 1;
+  AWS = 2;
+  GCP = 3;
+  AZURE = 4;
+}
+```
+
+When ORC is using the Hadoop or Ranger KMS, it generates a random encrypted
+local key (16 or 32 bytes for 128 or 256 bit AES respectively). Using the
+first 16 bytes as the IV, it uses AES/CTR to decrypt the local key.
+
+With the AWS KMS, the GenerateDataKey method is used to create a new local
+key and the Decrypt method is used to decrypt it.
+
+## Data Masks
+
+The user's data is statically masked before writing the unencrypted
+variant. Because the masking was done statically when the file was
+written, the information about the masking is just informational.
+
+The three standard masks are:
+
+* nullify - all values become null
+* redact - replace characters with constants such as X or 9
+* sha256 - replace string with the SHA 256 of the value
+
+The default is nullify, but masks may be defined by the user. Masks
+are not allowed to change the type of the column, just the values.
+
+```
+message DataMask {
+  // the kind of masking, which may include third party masks
+  optional string name = 1;
+  // parameters for the mask
+  repeated string maskParameters = 2;
+  // the unencrypted column roots this mask was applied to
+  repeated uint32 columns = 3 [packed = true];
+}
+```
+
+## Encryption Keys
+
+In addition to the encrypted local keys, which are stored in the
+footer's StripeInformation, the file also needs to describe the master
+key that was used to encrypt the local keys. The master keys are
+described by name, their version, and the encryption algorithm.
+
+```
+message EncryptionKey {
+  optional string keyName = 1;
+  optional uint32 keyVersion = 2;
+  optional EncryptionAlgorithm algorithm = 3;
+}
+```
+
+The encryption algorithm is stored using an enumeration and since
+ProtoBuf uses the 0 value as a default, we added an unused value. That
+ensures that if we add a new algorithm that old readers will get
+UNKNOWN_ENCRYPTION instead of a real value.
+
+```
+enum EncryptionAlgorithm {
+  // used for detecting future algorithms
+  UNKNOWN_ENCRYPTION = 0;
+  // 128 bit AES/CTR
+  AES_128 = 1;
+  // 256 bit AES/CTR
+  AES_256 = 2;
+}
+```
+
+## Encryption Variants
+
+Each encrypted column is written as two variants:
+
+* encrypted unmasked - for users with access to the key
+* unencrypted masked - for all other users
+
+The changes to the format were done so that old ORC readers will read
+the masked unencrypted data. Encryption variants encrypt a subtree of
+columns and use a single local key. The initial version of encryption
+support only allows the two variants, but this may be extended later
+and thus readers should use the first variant of a column that the
+reader has access to.
```

Comment:
The use case would be where you needed multiple levels of access. For example, if you wanted some of your data scientists to be able to see the actual data, others were only allowed to see the sha256 so that they could join against other tables, but general users couldn't see the data at all.

| Key | data |
| ---| ----- |
| pii | actual data |
| sha-pii | sha256 of the data |
| none | nulls |

Reply:
I am not sure if we should be doing this at Orc level. The scenario mentioned here can also be achieved from the application level by adding an additional column. That way user has much more flexibility to choose how they would expose information at different security level. Multiple encryption variants complicate the reader/writer quite a bit IMHO.

Reply:
I'm not proposing adding this in the initial release of column encryption, but I'd like to leave the possibility open. Given that we already have 2 variants, having overlapping encryption variants is very straightforward.

Reply:
Additionally, with users that I've talked to about this feature, some of them could really use it. 

Note that Hive & Ranger together provide this dynamically, if the user is only using Hive Server 2.

Reply:
When the reader starts planning a stripe read, it will try to decrypt the local keys. If it has access to the key for a given variant, it will use the streams & encodings for all of the columns under that variant's root column. So if the schema looks like: struct<a:int,b:struct<ba:string,bb:int>,c:int> , with "b" being encrypted. If the reader doesn't have access to the key, they'll just use the streams & encodings from the StripeFooter. If it does have access to the key, it will use the streams & encodings for "a", and "c" from the StripeFooter and "b", "ba", and "bb" from the StripeEncryptionVariant. Because the streams are separated by column, it is easy to do that filtering.

Let's get the first version released with column encryption and build more experience using it in production. That will help us understand what will help users the most.